src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java

```diff
@@ -6459,190 +6465,192 @@ void onFedAuthInfo(SqlFedAuthInfo fedAuthInfo, TDSTokenHandler tdsTokenHandler)
     }
 
     private SqlAuthenticationToken getFedAuthToken(SqlFedAuthInfo fedAuthInfo) throws SQLServerException {
-        // fedAuthInfo should not be null.
-        assert null != fedAuthInfo;
-
-        String user = activeConnectionProperties.getProperty(SQLServerDriverStringProperty.USER.toString());
-
-        // No of milliseconds to sleep for the initial back off.
-        int fedauthSleepInterval = BACKOFF_INTERVAL;
-
-        if (!msalContextExists()
-                && !authenticationString.equalsIgnoreCase(SqlAuthentication.ACTIVE_DIRECTORY_INTEGRATED.toString())) {
-            MessageFormat form = new MessageFormat(SQLServerException.getErrString("R_MSALMissing"));
-            throw new SQLServerException(form.format(new Object[] {authenticationString}), null, 0, null);
-        }
-
-        if (loggerExternal.isLoggable(java.util.logging.Level.FINEST)) {
-            loggerExternal.finest("Getting FedAuth token " + fedAuthInfo.toString());
-        }
-
-        while (true) {
-            int millisecondsRemaining = timerRemaining(timerExpire);
-            if (authenticationString.equalsIgnoreCase(SqlAuthentication.ACTIVE_DIRECTORY_PASSWORD.toString())) {
-                fedAuthToken = SQLServerMSAL4JUtils.getSqlFedAuthToken(fedAuthInfo, user,
-                        activeConnectionProperties.getProperty(SQLServerDriverStringProperty.PASSWORD.toString()),
-                        authenticationString, millisecondsRemaining);
-
-                // Break out of the retry loop in successful case.
-                break;
-            } else if (authenticationString
-                    .equalsIgnoreCase(SqlAuthentication.ACTIVE_DIRECTORY_MANAGED_IDENTITY.toString())) {
-
-                String managedIdentityClientId = activeConnectionProperties
-                        .getProperty(SQLServerDriverStringProperty.USER.toString());
-
-                if (null != managedIdentityClientId && !managedIdentityClientId.isEmpty()) {
+        try (PerformanceLog.Scope fedAuthScope = PerformanceLog.createScope(PerformanceLog.perfLoggerConnection, traceID, PerformanceActivity.TOKEN_ACQUISITION)) {
```

Comment:
I would actually put this in onFedAuthInfo, so that the scope includes calls to any access token callbacks (but not the last two lines in there that send the token to the server).